src/CoreWCF.Primitives/src/CoreWCF/Channels/MessageEncoderCompressionHandler.cs

```diff
@@ -13,45 +17,30 @@ internal static class MessageEncoderCompressionHandler
         internal const string DeflateContentEncoding = "deflate";
         private const int DecompressBlockSize = 1024;
 
-        internal static void DecompressBuffer(ref ArraySegment<byte> buffer, BufferManager bufferManager, CompressionFormat compressionFormat, long maxReceivedMessageSize)
+        internal static async ValueTask<ReadOnlySequence<byte>> DecompressBufferAsync(ReadOnlySequence<byte> buffer, CompressionFormat compressionFormat, long maxReceivedMessageSize)
         {
-            MemoryStream memoryStream = new MemoryStream(buffer.Array, buffer.Offset, buffer.Count);
-            int maxDecompressedSize = (int)Math.Min(maxReceivedMessageSize, int.MaxValue);
+            if (buffer.Length > maxReceivedMessageSize)
+            {
+                throw Fx.Exception.AsError(new QuotaExceededException(SR.Format(SRCommon.MaxReceivedMessageSizeExceeded, maxReceivedMessageSize)));
+            }
 
-            using (BufferManagerOutputStream bufferedOutStream = new BufferManagerOutputStream(SRCommon.MaxReceivedMessageSizeExceeded, 1024, maxDecompressedSize, bufferManager))
+            Pipe pipe = new();
+            foreach (ReadOnlyMemory<byte> memory in buffer)
```

Comment:
Using a pipe creates an additional buffer that you're effectively copying everything into just to create a PipeReader which you call AsStream against to pass to the compression streams. Just use PipeReader.Create(ReadOnlySequence).AsStream() and skip creating a Pipe and writing into it.

Reply:
done